### ISSUE TITLE
Order of operations in decoder layer

### ISSUE THREAD

**Issue by hsm207** [affinelayer/pix2pix-tensorflow#108]
In the paper, the U-Net decoder is described in terms of CDk, where CDk denotes a Convolution-BatchNorm-Dropout-ReLU layer.

In the code, it looks like the order of operations is ReLU-Convolution-BatchNorm-Dropout:

```
    for decoder_layer, (out_channels, dropout) in enumerate(layer_specs):
        skip_layer = num_encoder_layers - decoder_layer - 1
        with tf.variable_scope("decoder_%d" % (skip_layer + 1)):
            if decoder_layer == 0:
                # first decoder layer doesn't have skip connections
                # since it is directly connected to the skip_layer
                input = layers[-1]
            else:
                input = tf.concat([layers[-1], layers[skip_layer]], axis=3)

            rectified = tf.nn.relu(input)
            # [batch, in_height, in_width, in_channels] => [batch, in_height*2, in_width*2, out_channels]
            output = gen_deconv(rectified, out_channels)
            output = batchnorm(output)

            if dropout > 0.0:
                output = tf.nn.dropout(output, keep_prob=1 - dropout)

            layers.append(output)

    # decoder_1: [batch, 128, 128, ngf * 2] => [batch, 256, 256, generator_outputs_channels]
    with tf.variable_scope("decoder_1"):
        input = tf.concat([layers[-1], layers[0]], axis=3)
        rectified = tf.nn.relu(input)
        output = gen_deconv(rectified, generator_outputs_channels)
        output = tf.tanh(output)
        layers.append(output)
```

**Comment by nicolastah** [comment]
Hey there, I've asked myself the same question. The order of the operations **really** is: ` Convolution - BatchNorm - Dropout - ReLU layer`

I found out that the variable `layers` (for the _generator_) contains features maps **before** the **activation function.**  Let's have a look at  `encoder_1`:

```
 with tf.variable_scope("encoder_1"):
        output = gen_conv(generator_inputs, a.ngf, a)
        layers.append(output)
```
The features maps are `append` to `layers`  right _after_ convolution (batch norm is not applied to the first _C64_) Thus, when we fetch the previous layer `layers[-1]`,  we **first** need to apply the activation function:  

```
for out_channels in layer_specs:
  with tf.variable_scope("encoder_%d" % (len(layers) + 1)):
    rectified = lrelu(layers[-1], 0.2)
    convolved = gen_conv(rectified, out_channels, a)
    output = batchnorm(convolved)
    layers.append(output)
```
The _last encoder_ layer before the _first decoder_ layer did not go through an activation function. It is the reason why we start by applying the relu:

```
with tf.variable_scope("decoder_%d" % (skip_layer + 1)):
  if decoder_layer == 0:
    input = layers[-1]
  else:
    input = tf.concat([layers[-1], layers[skip_layer]], axis=3)

  rectified = tf.nn.relu(input)
  output = gen_deconv(rectified, out_channels, a)
  output = batchnorm(output)
  ...
```

- For the **generator,** `layers` represents  `Convolution - BatchNorm `
- For the **discriminator,** `layers` represents  `Convolution - BatchNorm - ReLU layer`
